File: src/Tools/DinoforgeMcp/dinoforge_mcp/native_dep_resolver.py
"""Unified resolver for native binary + game-path discovery (Python side).

Walks: env var -> installer-shipped paths.json -> hardcoded fallbacks -> raises.

Mirror of ``src/Tools/McpServer/NativeDepResolver.cs``. Same shape, same
``paths.json`` schema (logical key -> absolute path) so the C# and Python
sides agree on what the installer publishes.

Installer paths file location:
    Windows: %ProgramData%\\DINOForge\\paths.json
    Linux/macOS: /etc/dinoforge/paths.json
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Iterable, Optional


INSTALLER_PATHS_FILENAME = "paths.json"
INSTALLER_SUBDIR = "DINOForge"


class NativeDepNotFound(FileNotFoundError):
    """Raised when no candidate path resolves to an existing file/directory."""


def _installer_paths_file() -> Path:
    """Returns the absolute path to the installer's paths.json (regardless of existence)."""
    if os.name == "nt":
        program_data = os.environ.get("ProgramData", r"C:\ProgramData")
        return Path(program_data) / INSTALLER_SUBDIR / INSTALLER_PATHS_FILENAME
    return Path("/etc") / "dinoforge" / INSTALLER_PATHS_FILENAME


def _try_load_installer_paths() -> Optional[dict]:
    p = _installer_paths_file()
    if not p.exists():
        return None
    try:
        return json.loads(p.read_text(encoding="utf-8"))
    except Exception:
        return None


def _exists(path: str, require_file: bool) -> bool:
    p = Path(path)
    return p.is_file() if require_file else (p.is_file() or p.is_dir())


def resolve(
    key: str,
    env_var: str,
    hardcoded_fallbacks: Iterable[str],
    description: str,
    require_file: bool = True,
) -> str:
    """Resolve a native dependency. Returns absolute path. Raises NativeDepNotFound.

    Args:
        key: Logical key in installer paths.json (e.g. "playcua_native").
        env_var: Environment variable name (e.g. "PLAYCUA_NATIVE_EXE").
        hardcoded_fallbacks: Last-resort paths tried in order.
        description: Human-readable description used in the error message.
        require_file: If True (default) candidates must be files; if False, dirs are accepted.
    """
    fallbacks = list(hardcoded_fallbacks)

    # 1. Env var
    from_env = os.environ.get(env_var)
    if from_env and _exists(from_env, require_file):
        return from_env

    # 2. Installer paths.json
    installer_paths = _try_load_installer_paths()
    from_installer: Optional[str] = None
    if installer_paths and key in installer_paths:
        from_installer = installer_paths[key]
        if from_installer and _exists(from_installer, require_file):
            return from_installer

    # 3. Hardcoded fallbacks
    for fb in fallbacks:
        if fb and _exists(fb, require_file):
            return fb

    # 4. Loud error
    raise NativeDepNotFound(
        f"Native dependency {description!r} (key={key}) not found.\n"
        f"  Env ${env_var}: {from_env or '<unset>'}\n"
        f"  Installer paths.json key: {from_installer or '<not in paths.json>'}\n"
        f"  Hardcoded fallbacks: {fallbacks}\n"
        f"Set ${env_var} or run the DINOForge installer to populate paths.json."
    )


def try_resolve(
    key: str,
    env_var: str,
    hardcoded_fallbacks: Iterable[str],
    require_file: bool = True,
) -> Optional[str]:
    """Try-resolve variant. Returns None instead of raising when nothing matches."""
    try:
        return resolve(key, env_var, hardcoded_fallbacks, key, require_file)
    except NativeDepNotFound:
        return None

File: src/Tools/McpServer/NativeDepResolver.cs
#nullable enable
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;

namespace DINOForge.Tools.McpServer;

/// <summary>
/// Unified resolver for native binary + game-path discovery.
/// Walks: env var -> installer-shipped paths.json -> hardcoded fallback -> throws.
/// </summary>
/// <remarks>
/// Replaces three ad-hoc sites that each had their own bespoke fallback chain:
///   - GameCaptureHelper.ResolveBepInExRoot (DINO game path)
///   - GameCaptureHelper.ResolveBareCuaPath (bare-cua-native)
///   - isolation_layer.PlayCUABackend._resolve_playcua_path (Python sibling, see native_dep_resolver.py)
///
/// The installer (when shipped) writes a JSON dictionary to
/// %ProgramData%\DINOForge\paths.json on Windows or /etc/dinoforge/paths.json on Linux,
/// mapping logical keys (e.g. "dino_game_path") to absolute paths discovered at install time.
/// </remarks>
public static class NativeDepResolver
{
    private const string InstallerPathsFile = "paths.json";
    private const string InstallerSubdir = "DINOForge";

    /// <summary>
    /// Resolve a native dependency by key.
    /// </summary>
    /// <param name="key">Logical key in installer paths.json (e.g. "dino_game_path", "bare_cua_native", "playcua_native").</param>
    /// <param name="envVar">Environment variable name that overrides everything (e.g. "DINOFORGE_GAME_PATH").</param>
    /// <param name="hardcodedFallbacks">Last-resort paths tried in order. Used for dev convenience only.</param>
    /// <param name="description">Human-readable description used in the error message.</param>
    /// <param name="requireFile">If true, candidate must be a file. If false (game-path mode), a directory is acceptable.</param>
    /// <returns>An absolute path that exists.</returns>
    /// <exception cref="FileNotFoundException">Thrown when no candidate exists. The message lists every probe attempt.</exception>
    public static string Resolve(
        string key,
        string envVar,
        IEnumerable<string> hardcodedFallbacks,
        string description,
        bool requireFile = true)
    {
        var fallbacks = hardcodedFallbacks?.ToArray() ?? Array.Empty<string>();

        // 1. Env var
        var fromEnv = Environment.GetEnvironmentVariable(envVar);
        if (!string.IsNullOrWhiteSpace(fromEnv) && Exists(fromEnv, requireFile))
            return fromEnv;

        // 2. Installer paths.json
        var installerPaths = TryLoadInstallerPaths();
        string? fromInstaller = null;
        if (installerPaths != null && installerPaths.TryGetValue(key, out var installerVal))
        {
            fromInstaller = installerVal;
            if (!string.IsNullOrWhiteSpace(fromInstaller) && Exists(fromInstaller, requireFile))
                return fromInstaller;
        }

        // 3. Hardcoded fallbacks
        string? fromFallback = fallbacks.FirstOrDefault(f => !string.IsNullOrWhiteSpace(f) && Exists(f, requireFile));
        if (fromFallback is not null)
            return fromFallback;

        // 4. Loud error
        var fallbackList = fallbacks.Length > 0 ? string.Join(", ", fallbacks) : "<none>";
        throw new FileNotFoundException(
            $"Native dependency '{description}' (key={key}) not found. Tried:\n" +
            $"  - Env var ${envVar}: {(string.IsNullOrEmpty(fromEnv) ? "<unset>" : fromEnv)}\n" +
            $"  - Installer paths.json: {fromInstaller ?? "<not in paths.json>"}\n" +
            $"  - Hardcoded fallbacks: [{fallbackList}]\n" +
            $"Set ${envVar} to override, or run the DINOForge installer to populate paths.json."
        );
    }

    /// <summary>
    /// Try-resolve variant. Returns null instead of throwing when nothing matches.
    /// Useful for "best-effort" call sites (e.g. optional fallback binaries).
    /// </summary>
    public static string? TryResolve(
        string key,
        string envVar,
        IEnumerable<string> hardcodedFallbacks,
        bool requireFile = true)
    {
        try
        {
            return Resolve(key, envVar, hardcodedFallbacks, key, requireFile);
        }
        catch (FileNotFoundException)
        {
            return null;
        }
    }

    private static bool Exists(string path, bool requireFile)
    {
        return requireFile ? File.Exists(path) : (File.Exists(path) || Directory.Exists(path));
    }

    /// <summary>
    /// Locates and parses the installer paths.json. Returns null when the file
    /// is missing, unreadable, or malformed (the resolver continues to fallbacks).
    /// </summary>
    internal static string GetInstallerPathsFilePath()
    {
        var programData = Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData);
        if (string.IsNullOrEmpty(programData))
            programData = OperatingSystem.IsWindows() ? @"C:\ProgramData" : "/etc";
        return Path.Combine(programData, InstallerSubdir, InstallerPathsFile);
    }

    private static Dictionary<string, string>? TryLoadInstallerPaths()
    {
        var pathsFile = GetInstallerPathsFilePath();
        if (!File.Exists(pathsFile)) return null;

        try
        {
            var json = File.ReadAllText(pathsFile, Encoding.UTF8);
            return JsonSerializer.Deserialize<Dictionary<string, string>>(json);
        }
        catch
        {
            return null;
        }
    }
}
